tutorials/cyber/network_mapping/lanl_network_mapping_using_rapids.ipynb

```diff
@@ -0,0 +1,801 @@
+    "                        'account_for_which_logon_failed_account_name',\n",
+    "                        'network_information_workstation_name',\n",
+    "                        'network_information_source_port',\n",
+    "                        'subject_security_id']\n",
+    "OUTPUT_DTYPES = ['str' for x in OUTPUT_COLS_SUPERSET]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Parsing Data Pipeline\n",
+    "\n",
+    "Preprocess, or clean the raw logs, by removing the non-printable characters then begin to parse logs by event code type. Here we use regex mappings per event code to parse out each data element defined in OUTPUT_COLS_SUPERSET"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def pipeline(df, event_codes, regex_mappings, clean=True):\n",
+    "    \"\"\"\n",
+    "    \"\"\"\n",
+    "    \n",
+    "    if clean:\n",
+    "        df = preprocess_logs(df)\n",
+    "    out_dfs = [] \n",
+    "    \n",
+    "    # separate by eventcode and process differently\n",
+    "    \n",
+    "    for code in event_codes:\n",
+    "        portion = filter_by_pattern(df, code)\n",
+    "        temp = process_logtype(portion, regex_mappings, code)\n",
+    "        temp['time'] = portion['Time'].astype('int')\n",
+    "        temp['eventcode'] = portion['EventID']\n",
+    "        out_dfs.append(temp)\n",
+    "        \n",
+    "    # recombine the processed output\n",
+    "    out_df = cudf.multi.concat(out_dfs)\n",
+    "    return out_df\n",
+    "\n",
+    "\n",
+    "def concat_wrapper(df_list):\n",
+    "    return cudf.multi.concat(df_list)\n",
+    "\n",
+    "\n",
+    "def preprocess_logs(logs_gdf):\n",
+    "    \"\"\"Lowercasing and replacing characters\n",
+    "    \"\"\"\n",
+    "    logs_gdf['Raw'] = (\n",
+    "        logs_gdf['Raw'].str.lower()\n",
+    "        .str.replace('\\\\\\\\t', '')\n",
+    "        .str.replace('\\\\\\\\r', '')\n",
+    "        .str.replace('\\\\\\\\n', ' | ')\n",
+    "    )\n",
+    "    return logs_gdf\n",
+    "\n",
+    "\n",
+    "def process_logtype(df, regexes, eventcode):\n",
+    "    \"\"\"Ongoing strings development/fixes will allow for cleaner log processing code in the future\n",
+    "    \"\"\"\n",
+    "\n",
+    "    log_df_processed = cudf.read_csv('conf/LANL_OUTPUT_COLS_SUPERSET.csv', dtype=OUTPUT_DTYPES)\n",
+    "    log_df_processed = log_df_processed[:0]\n",
+    "    \n",
+    "    columns = list(regexes[eventcode].keys())\n",
+    "\n",
+    "    for col in columns:\n",
+    "        regex_pattern = regexes[eventcode].get(col)\n",
+    "        extracted_nvstrings = df['Raw'].str.extract(regex_pattern)\n",
+    "        \n",
+    "        if not extracted_nvstrings.empty:\n",
+    "            log_df_processed[col] = extracted_nvstrings[0]\n",
+    "    \n",
+    "    for col in log_df_processed.columns:\n",
```

Comment:
I'm not 100% sure without testing, but I believe some of this hack-y code is no longer necessary with some fixes to the dtype handling of adding new columns to dataframes. Worth taking a look at whether this can be simplified.